logbook-jdkserver/src/main/java/org/zalando/logbook/jdkserver/Request.java

```diff
@@ -0,0 +1,231 @@
+
+        Buffering(final byte[] body) {
+            this(body, new ByteArrayInputStream(body));
+        }
+
+        Buffering(final byte[] body, final ByteArrayInputStream stream) {
+            super(stream);
+            this.body = body;
+        }
+
+        @Override
+        public State without() {
+            return new Ignoring(body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    private static final class Ignoring extends Streaming {
+
+        private final byte[] body;
+
+        Ignoring(final byte[] body) {
+            super(new ByteArrayInputStream(body));
+            this.body = body;
+        }
+
+        @Override
+        public State with() {
+            return new Buffering(body, getStream());
+        }
+
+    }
+
+    public Request(HttpExchange httpExchange) {
+        this.httpExchange = httpExchange;
+        this.state = new AtomicReference<>(new Unbuffered());
+    }
+
+    @Override
+    public Origin getOrigin() {
+        return Origin.REMOTE;
+    }
+
+    @Override
+    public HttpHeaders getHeaders() {
+        HttpHeaders headers = HttpHeaders.empty();
+        for (Map.Entry<String, List<String>> e : httpExchange.getRequestHeaders().entrySet()) {
+            headers = headers.update(e.getKey(), e.getValue());
+        }
+        return headers;
+    }
+
+    @Override
+    public String getRemote() {
+        return httpExchange.getRemoteAddress().getHostString();
+    }
+
+    @Override
+    public String getMethod() {
+        return httpExchange.getRequestMethod();
+    }
+
+    @Override
+    public String getScheme() {
+        String scheme = httpExchange.getRequestURI().getScheme();
+        if (scheme == null) {
+            scheme = "http";
+        }
+        return scheme;
+    }
+
+    @Override
+    public String getHost() {
+        return httpExchange.getLocalAddress().getHostString();
+    }
+
+    @Override
+    public Optional<Integer> getPort() {
+        return Optional.of(httpExchange.getLocalAddress().getPort());
+    }
+
+    @Override
+    public String getPath() {
+        return notNull(httpExchange.getRequestURI().getPath());
+    }
+
+    @Override
+    public String getQuery() {
+        return notNull(httpExchange.getRequestURI().getQuery());
+    }
+
+    @Override
+    public HttpRequest withBody() {
+        state.updateAndGet(State::with);
+        return this;
+    }
+
+    @Override
+    public HttpRequest withoutBody() {
+        state.updateAndGet(State::without);
+        return this;
+    }
+
+    public InputStream getInputStream() throws IOException {
+        return buffer().getInputStream(httpExchange);
+    }
+
+    @Override
+    public byte[] getBody() {
+        return buffer().getBody();
+    }
+
+    @Override
+    public String getProtocolVersion() {
+        return httpExchange.getProtocol();
+    }
+
+    @Override
+    public String getRequestUri() {
```

Comment:
Yes, I think it could be used but isn't this better then decomposing and then reconstructing it while we have it directly available? I though it is intended like that - having the default while override it with better option if available.

Reply:
Fair enough